cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go

```diff
@@ -70,34 +83,80 @@ func (o *checkCapacityProvClass) Provision(
 	daemonSets []*appsv1.DaemonSet,
 	nodeInfos map[string]*schedulerframework.NodeInfo,
 ) (*status.ScaleUpStatus, errors.AutoscalerError) {
-	if len(unschedulablePods) == 0 {
-		return &status.ScaleUpStatus{Result: status.ScaleUpNotTried}, nil
-	}
+	combinedStatus := NewCombinedStatusSet()
+	provisioningRequestsProcessed := 0
+	startTime := time.Now()
 
-	prs := provreqclient.ProvisioningRequestsForPods(o.client, unschedulablePods)
-	prs = provreqclient.FilterOutProvisioningClass(prs, v1.ProvisioningClassCheckCapacity)
-	if len(prs) == 0 {
-		return &status.ScaleUpStatus{Result: status.ScaleUpNotTried}, nil
-	}
-	// Pick 1 ProvisioningRequest.
-	pr := prs[0]
 	o.context.ClusterSnapshot.Fork()
 	defer o.context.ClusterSnapshot.Revert()
 
-	scaleUpIsSuccessful, err := o.checkcapacity(unschedulablePods, pr)
-	if err != nil {
-		return status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
-	}
-	if scaleUpIsSuccessful {
-		return &status.ScaleUpStatus{Result: status.ScaleUpSuccessful}, nil
+	for len(unschedulablePods) > 0 {
+		prs := provreqclient.ProvisioningRequestsForPods(o.client, unschedulablePods)
+		prs = provreqclient.FilterOutProvisioningClass(prs, v1.ProvisioningClassCheckCapacity)
+		if len(prs) == 0 {
+			break
+		}
+
+		// Pick 1 ProvisioningRequest.
+		pr := prs[0]
+
+		scaleUpIsSuccessful, err := o.checkcapacity(unschedulablePods, pr)
+		if err != nil {
+			st, err := status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
+
+			if o.checkCapacityBatchProcessing {
+				combinedStatus.Add(st)
+			} else {
+				return st, err
+			}
+		}
+		if scaleUpIsSuccessful {
+			combinedStatus.Add(&status.ScaleUpStatus{Result: status.ScaleUpSuccessful})
+		} else {
+			combinedStatus.Add(&status.ScaleUpStatus{Result: status.ScaleUpNoOptionsAvailable})
+		}
+
+		if !o.checkCapacityBatchProcessing {
+			break
+		}
+
+		if o.provreqInjector == nil {
+			klog.Errorf("ProvisioningRequestPodsInjector is not set, falling back to non-batch processing")
+			break
+		}
+
+		if o.checkCapacityProvisioningRequestMaxBatchSize <= 1 {
+			klog.Errorf("MaxBatchSize is set to %d, falling back to non-batch processing", o.checkCapacityProvisioningRequestMaxBatchSize)
+			break
+		}
+
+		provisioningRequestsProcessed++
+		if provisioningRequestsProcessed >= o.checkCapacityProvisioningRequestMaxBatchSize {
+			break
+		}
+
+		if time.Since(startTime) > o.checkCapacityProvisioningRequestBatchTimebox {
+			klog.Infof("Batch timebox exceeded, processed %d check capacity provisioning requests this iteration", provisioningRequestsProcessed)
+			break
+		}
+
+		unschedulablePods, err = (*o.provreqInjector).GetPodsFromNextRequest(func(pr *provreqwrapper.ProvisioningRequest) bool {
```

Comment:
An implementation of this optimization where all pending PRs are injected is planned and will be implemented in the near future.

Reply:
Ack, I'd appreciate leaving a TODO about the planned rework.

If we're leaving the current flow, can we make this loop more readable? Right now it's roughly:
```
unschedulable_pods := ...
if len(unschedulable_pods) > 0 {
  pr := podsToPr(unschedulablePods)
  err := checkCapacity(unschedulable_pods, pr)
  if err != nil {
    // This part is the most unclear to me right now.
  }
  if should_stop_processing() {
    break
  }
  unschedulable_pods := get_next_unschedulable_pods()
}
```

Extracting these helper methods would make it way more readable IMO. It might also read a bit better if we're not going through `unschedulablePods`:

```
unschedulable_pods := ...

pr_pods := unschedulablePods
for len(pr_pods) > 0 {
  pr := podsToPr(pr_pods)
  err := checkCapacity(pr_pods, pr)
  if err != nil {
    ...
  }
  if should_stop_processing() {
    break
  }
  pr_pods  = get_next_pr_pods()
}
```

Reply:
I added a TODO comment and extracted some logic to helper functions.
>  This part is the most unclear to me right now.
I simplified this part. The intention is that if the combined status has any PR with a status of a higher priority than ScaleUpError, then an error will not be returned as the second return value.